pkg/streamingpromql/operators/functions/label.go

```diff
@@ -17,6 +18,43 @@ import (
 	"github.com/grafana/mimir/pkg/streamingpromql/types"
 )
 
+func LabelJoinFactory(dstLabelOp, separatorOp types.StringOperator, srcLabelOp []types.StringOperator) SeriesMetadataFunction {
+	return func(seriesMetadata []types.SeriesMetadata, _ *limiting.MemoryConsumptionTracker) ([]types.SeriesMetadata, error) {
+		dst := dstLabelOp.GetValue()
+		if !model.LabelName(dst).IsValid() {
+			return nil, fmt.Errorf("invalid destination label name in label_join(): %s", dst)
+		}
+		separator := separatorOp.GetValue()
+		srcLabels := make([]string, len(srcLabelOp))
+		for i, op := range srcLabelOp {
+			src := op.GetValue()
+			if !model.LabelName(src).IsValid() {
+				return nil, fmt.Errorf("invalid source label name in label_join(): %s", dst)
+			}
+			srcLabels[i] = src
+		}
+
+		lb := labels.NewBuilder(labels.EmptyLabels())
+		labelValues := make([]string, 0, len(srcLabels))
+
+		for i := range seriesMetadata {
+			labelValues = labelValues[:0]
+
+			for _, srcLabel := range srcLabels {
+				// Get returns an empty string for missing labels, so this is safe and gives the desired output
+				// where a series may be missing a source label.
+				labelValues = append(labelValues, seriesMetadata[i].Labels.Get(srcLabel))
+			}
+
+			lb.Reset(seriesMetadata[i].Labels)
+			lb.Set(dst, strings.Join(labelValues, separator))
```

Comment:
It'd be interesting to see if using the same `strings.Builder` here for every series would be more efficient - `strings.Join` uses a new `strings.Builder` under the covers for each call, so reusing the same one ourselves should allow us to avoid allocating a new buffer for every series.